packages/lit-html/src/parts.ts

```diff
@@ -0,0 +1,117 @@
+import {NodePart, nothing} from './lit-html.js';
+
+/**
+ * The state of a NodePart, which can be detached and reattached.
+ */
+export type NodePartState = {};
+
+/**
+ * The private interface for NodePartState, which should be kept opaque.
+ */
+type NodePartStateInternal = {
+  __value: unknown;
+  __fragment: DocumentFragment;
+};
+
+export const detachNodePart = (part: NodePart): NodePartState => {
+  const fragment = document.createDocumentFragment();
+  const state: NodePartStateInternal = {
+    __value: part.__value,
+    __fragment: fragment,
+  };
+  let start = part.__startNode.nextSibling;
+  let nextNode;
+  while (start !== part.__endNode) {
+    nextNode = start!.nextSibling;
+    fragment.append(start!);
+    start = nextNode;
+  }
+  part.__value = nothing;
+  return state;
+};
+
+export const restoreNodePart = (part: NodePart, state: NodePartState) => {
+  // TODO (justinfagnani): make an interal-only interface
+  (part as any).__commitNode((state as NodePartStateInternal).__fragment);
+  part.__value = (state as NodePartStateInternal).__value;
+};
+
+const createMarker = () => document.createComment('');
+
+export const createAndInsertPart = (
+  containerPart: NodePart,
+  refPart?: NodePart
+): NodePart => {
+  const container = containerPart.__startNode.parentNode as Node;
+
+  const endNode =
+    refPart === undefined ? containerPart.__endNode : refPart.__startNode;
+
+  const startNode = container.insertBefore(createMarker(), endNode);
+
+  container.insertBefore(createMarker(), endNode);
+  const newPart = new NodePart(startNode, endNode, containerPart.options);
+  return newPart;
+};
+
+export const updatePart = (part: NodePart, value: unknown) => {
```

Comment:
Just noting that from an API point of view, this method is pretty marginal; it made a little more sense in 1.x when it was set+commit (and was useful to return the part so repeat could do update+assign as one-liners, but that's really just a convenience for repeat).

I guess its real value is giving a public API for `__setValue` _just_ for `NodePart` (which we have valid use cases for); and the idea is we would not need an analog for `AttributePart` (whose `setValue` API is more complex) because you shouldn't be constructing those yourself, and thus so returning the value from `render` is sufficient?

Reply:
But... we already differentiate the `__setValue` call between Attribute & NodeParts... so seems like NodePart's setValue could just be public?  Something to consider.

Reply:
Yeah, this would let us still minify the Part API. Let's consider as part of the Part API overall.